README.rst
=====================
django-specifications
=====================

.. image:: https://travis-ci.org/matthiask/django-specifications.png?branch=master
   :target: https://travis-ci.org/matthiask/django-specifications

This module offers an easy way to attach auxiliary information to Django
models. It allows configuring custom forms through the administration
interface.

Usage
=====

1. Add ``"specifications"`` to ``INSTALLED_APPS``.
2. Create a ``specification = ForeignKey("specifications.Specification")``
   on the model you want to use specifications with. The foreign key can
   be nullable or required, as you wish.
3. Create the place where the specification field data is actually stored::

       from specifications.models import SpecificationValueFieldBase

       class MyObjectField(SpecificationValueFieldBase):
           parent = models.ForeignKey(
               MyObject,
               on_delete=models.CASCADE,
               related_name="fields",
           )

           class Meta:
               ordering = ["field__group__ordering", "ordering"]

4. Inherit from ``FormWithSpecification`` when creating your ``ModelForm``::

       from specifications.forms import FormWithSpecification

       class MyObjectForm(FormWithSpecification):
           class Meta:
               model = MyObject

5. If you want to edit models with specifications you might want to use
   the following snippet::

       from specifications.admin import ModelAdminWithSpecification

       class MyObjectAdmin(ModelAdminWithSpecification):
           pass

The fields are available after saving a specification. The specification can
be changed, but you risk losing data if you do this.


CHANGELOG.rst

Change log
==========

`Next version`_
~~~~~~~~~~~~~~~


`0.5`_ (2021-09-13)
~~~~~~~~~~~~~~~~~~~

- Dropped compatibility with Python < 3.8.
- Dropped compatibility with Django < 3.2.
- Modernized the package, added isort etc, moved from Travis CI to GitHub
  actions.


`0.4`_ (2019-09-13)
~~~~~~~~~~~~~~~~~~~

- Added a helper for adding specification field group fieldsets.
- Fixed a bug because the code wouldn't handle the case where several
  specification fields were gone at once.
- Dropped compatibility guarantees with old versions of Django.


`0.3`_ (2018-09-30)
~~~~~~~~~~~~~~~~~~~

- Reformatted the code using black.
- Added ``fields = '__all__'`` to the modelform as required by Django.
- Replaced the migrations generated by South.
- Added support for Django 1.8 LTS, 1.11 LTS and 2.1 while dropping
  support for some older versions. Added support for Python 3.
- Added a testsuite, Travis CI integration and documentation on
  readthedocs.io.
- Added a ``ModelAdminWithSpecification`` class which is especially
  useful if you want to group specification fields into their own
  fieldsets.


`0.2`_ (2014-10-06)
~~~~~~~~~~~~~~~~~~~

- Switched from distutils to setuptools.
- Added utilities for analyzing and comparing objects with
  specifications.


`0.1`_ (2012-12-18)
~~~~~~~~~~~~~~~~~~~

- Initial public version.

.. _0.1: https://github.com/matthiask/django-specifications/commit/efc41b6f5e4
.. _0.2: https://github.com/matthiask/django-specifications/compare/0.1...0.2
.. _0.3: https://github.com/matthiask/django-specifications/compare/0.2...0.3
.. _0.4: https://github.com/matthiask/django-specifications/compare/0.3...0.4
.. _0.5: https://github.com/matthiask/django-specifications/compare/0.4...0.5
.. _Next version: https://github.com/matthiask/django-specifications/compare/0.5...master


specifications/migrations/0001_initial.py
# Generated by Django 2.1.1 on 2018-09-29 07:56

import django.db.models.deletion
from django.db import migrations, models


class Migration(migrations.Migration):

    initial = True

    dependencies = []

    operations = [
        migrations.CreateModel(
            name="Specification",
            fields=[
                (
                    "id",
                    models.AutoField(
                        auto_created=True,
                        primary_key=True,
                        serialize=False,
                        verbose_name="ID",
                    ),
                ),
                ("name", models.CharField(max_length=100, verbose_name="name")),
                ("notes", models.TextField(blank=True, verbose_name="notes")),
            ],
            options={
                "verbose_name": "specification",
                "verbose_name_plural": "specifications",
                "ordering": ["name"],
            },
        ),
        migrations.CreateModel(
            name="SpecificationField",
            fields=[
                (
                    "id",
                    models.AutoField(
                        auto_created=True,
                        primary_key=True,
                        serialize=False,
                        verbose_name="ID",
                    ),
                ),
                ("name", models.CharField(max_length=100, verbose_name="name")),
                (
                    "type",
                    models.CharField(
                        choices=[
                            ("text", "text"),
                            ("longtext", "long text"),
                            ("boolean", "boolean"),
                            ("integer", "integer"),
                            ("closed_set_single", "closed set"),
                            ("closed_set_multiple", "closed set (multiple)"),
                            ("open_set_single", "open set"),
                            ("open_set_single_extensible", "open set (extensible)"),
                        ],
                        max_length=30,
                        verbose_name="type",
                    ),
                ),
                (
                    "choices",
                    models.TextField(
                        blank=True,
                        help_text="One choice per line (if applicable).",
                        verbose_name="choices",
                    ),
                ),
                (
                    "help_text",
                    models.CharField(
                        blank=True, default="", max_length=100, verbose_name="help text"
                    ),
                ),
                (
                    "required",
                    models.BooleanField(verbose_name="required", default=True),
                ),
                ("ordering", models.IntegerField(default=0, verbose_name="ordering")),
            ],
            options={
                "verbose_name": "specification field",
                "verbose_name_plural": "specification fields",
                "ordering": ["group__ordering", "ordering"],
            },
        ),
        migrations.CreateModel(
            name="SpecificationFieldGroup",
            fields=[
                (
                    "id",
                    models.AutoField(
                        auto_created=True,
                        primary_key=True,
                        serialize=False,
                        verbose_name="ID",
                    ),
                ),
                ("name", models.CharField(max_length=100, verbose_name="name")),
                ("ordering", models.IntegerField(default=0, verbose_name="ordering")),
                (
                    "specification",
                    models.ForeignKey(
                        on_delete=django.db.models.deletion.CASCADE,
                        related_name="groups",
                        to="specifications.Specification",
                        verbose_name="specification",
                    ),
                ),
            ],
            options={
                "verbose_name": "specification field group",
                "verbose_name_plural": "specification field groups",
                "ordering": ["ordering"],
            },
        ),
        migrations.AddField(
            model_name="specificationfield",
            name="group",
            field=models.ForeignKey(
                blank=True,
                null=True,
                on_delete=django.db.models.deletion.SET_NULL,
                related_name="fields",
                to="specifications.SpecificationFieldGroup",
                verbose_name="group",
            ),
        ),
        migrations.AddField(
            model_name="specificationfield",
            name="specification",
            field=models.ForeignKey(
                on_delete=django.db.models.deletion.CASCADE,
                related_name="fields",
                to="specifications.Specification",
                verbose_name="specification",
            ),
        ),
    ]
